libnetwork/portmapper/proxy_linux.go

```diff
@@ -1,78 +1,64 @@
 package portmapper
 
 import (
+	"errors"
 	"fmt"
 	"io"
-	"net"
 	"os"
 	"os/exec"
 	"runtime"
 	"strconv"
 	"syscall"
 	"time"
-)
 
-// StartProxy starts the proxy process at proxyPath, or instantiates a dummy proxy
-// to bind the host port if proxyPath is the empty string.
-func StartProxy(
-	proto string,
-	hostIP net.IP, hostPort int,
-	containerIP net.IP, containerPort int,
-	proxyPath string,
-) (stop func() error, retErr error) {
-	if proxyPath == "" {
-		return newDummyProxy(proto, hostIP, hostPort)
-	}
-	return newProxyCommand(proto, hostIP, hostPort, containerIP, containerPort, proxyPath)
-}
+	"github.com/docker/docker/libnetwork/types"
+)
 
-func newProxyCommand(
-	proto string,
-	hostIP net.IP, hostPort int,
-	containerIP net.IP, containerPort int,
+// StartProxy starts the proxy process at proxyPath.
+// If listenSock is not nil, it must be a bound socket that can be passed to
+// the proxy process for it to listen on. StartProxy always takes ownership
+// of listenSock if proxyPath is nonzero, the socket will be closed even if
+// the proxy fails to start.
```

Comment:
I don't think it is necessary for StartProxy to clobber listenSock. Even in the successful case where the open socket is passed into the docker-proxy subprocess, the file descriptor in the dockerd process remains a valid open reference to the underlying file description. Having StartProxy close the file descriptor is therefore _policy_, and you know what my opinions are on policy. Let the caller decide what to do with it. For example, having the caller retain ownership of the socket fd affords:
- retrying starting the proxy (after a short delay) if something goes wrong
- relaunching the proxy process with the same socket if it dies unexpectedly
- seamlessly upgrading the proxy on the fly (start new proxy process listening on socket, signal old proxy to stop accepting new connections and exit once the connection count drops to zero.)

Reply:
Done.